---
title: "Hierarchical Regulation of Autophagy During Adipocyte Differentiation"
author: "Mahmoud Shaaban"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
```

```{r load_libraries}
# load required libraries
library(segmentr)
library(tidyverse)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(org.Mm.eg.db)
library(ComplexHeatmap)
library(circlize)
library(SummarizedExperiment)
library(reshape2)
library(ggupset)
library(cowplot)
library(DESeq2)
library(ggupset)
library(xtable)
library(scales)
library(segmentr)
```

```{r autophagy_expression}
# load data
gene_annotations <- read_rds('data/gene_annotations.rds') %>%
    as_tibble() %>%
    dplyr::select(category, gene_id = SYMBOL)

# numbers of differentially expressed genes
differential_expression <- read_rds('data/differential_expression.rds') %>%
    left_join(gene_annotations) %>%
    filter(padj < .2) %>%
    mutate(dir = ifelse(fold_change > 0, 'Up', 'Down'),
           time = fct_reorder(factor(time), time)) %>%
    group_by(category, course, time, dir) %>%
    summarise(n = n())

gene_annotations <- read_rds('data/gene_annotations.rds') %>%
    with(split(SYMBOL, category)) %>%
    map(unique)

adipo_counts <- read_rds('data/adipo_counts.rds')
adipo_counts$time <- relevel(factor(adipo_counts$time), ref = '0')

dds <- DESeqDataSet(adipo_counts, ~time)
trans_counts <- vst(dds)

factors <- c('Pparg', 'Cebpb', 'Med1', 'EP300', 'Rxrg')

ind <- rownames(adipo_counts) %in% c(gene_annotations$Autophagy, factors)
mat <- assay(trans_counts)[ind,]

vars <- apply(mat, 1, var)
variance <- tibble(gene = rownames(mat),
                   var_exp = vars/sum(vars))

gene_set_enrichment <- read_rds('data/gene_set_enrichment.rds')

plot_grid(
    differential_expression %>%
    filter(category == 'Autophagy',
           course == 'None') %>%
    ggplot(aes(x = time, y = n, fill = dir)) +
    geom_col(position = 'dodge') +
    labs(x = 'Time (hr)',
         y = 'Number of Genes',
         fill = 'Regualtion') +
    lims(y = c(0, 90)) +
    scale_fill_manual(values = c('black', 'darkgray')) +
    theme(legend.position = c(.5, .9),
          legend.direction = 'horizontal',
          legend.background = element_blank(),
          panel.grid = element_blank(),
          axis.text = element_text(size = 11),
          axis.title = element_text(size = 14)),
        gene_set_enrichment %>%
    filter(grepl('None', group),
    pathway %in% c('GO:0006914', 'GO:0006629')) %>%
    separate(group, c('course', 'time'), '_') %>%
    mutate(time = fct_reorder(factor(time), as.numeric(time))) %>%
    ggplot(aes(x = time, y= NES, group = pathway)) +
    geom_point() +
    geom_line() +
    annotate('text', c(6.5, 6.4), y = c(0, 1.4),
             label = c('Autophagy', 'Lipid metabolic\n process')) +
    labs(x = 'Time (hr)',
         y = 'Enrichment Score') +
    theme(legend.position = 'none',
          panel.grid = element_blank(),
          axis.text = element_text(size = 11),
          axis.title = element_text(size = 14)),
    differential_expression %>%
    filter(category == 'Autophagy',
           course != 'None') %>%
    ggplot(aes(x = time, y = n, fill = dir)) +
    geom_col(position = 'dodge') +
    facet_grid(~course, scales = 'free_x', space = 'free_x') +
    lims(y = c(0, 40)) +
    scale_fill_manual(values = c('black', 'darkgray')) +
    labs(x = 'Time (hr)',
         y = 'Number of Genes',
         fill = 'Regualtion') +
    theme(legend.position = c(.5, .9),
          legend.direction = 'horizontal',
          legend.background = element_blank(),
          panel.grid = element_blank(),
          strip.background = element_blank(),
          axis.text = element_text(size = 12),
          axis.title = element_text(size = 14)),
    variance %>%
    filter(gene %in% factors) %>%
    mutate(gene = fct_reorder(gene, var_exp)) %>%
    ggplot(aes(x = gene, y = var_exp, .desc = TRUE)) +
    geom_col(width = .8) +
    labs(x = 'Factor Gene',
         y = 'Variance Explained') +
    theme(panel.grid = element_blank(),
          axis.text = element_text(size = 12),
          axis.title = element_text(size = 14)),
    nrow = 2,
    labels = 'AUTO',
    label_fontface = 'plain',
    scale = .95
) %>%
    ggsave(plot = .,
           filename = 'manuscript2/figures/autophagy_expression.png',
           height = 6, width = 6)
```

```{r autophagy_categories}
cats <- list(
    'Regulation' = list(
        'GO:0010507' = 'negative regulation of autophagy',
        'GO:0010508' = 'positive regulation of autophagy'
    ),
    'Selective' = list(
        'GO:0035973' = 'aggrephagy',
        'GO:0000423' = 'mitophagy',
        'GO:0061709' = 'reticulophagy',
        'GO:0098792' = 'xenophagy'
    ),
    'Subtypes' = list(
        'GO:0000422' = 'autophagy of mitochondrion',
        'GO:0030242' = 'autophagy of peroxisome',
        'GO:0061684' = 'chaperone-mediated autophagy',
        'GO:0061738' = 'late endosomal microautophagy',
        'GO:0016236' = 'macroautophagy')
) %>%
    melt() %>%
    setNames(c('subcat', 'pathway', 'cat'))

cats <- list(
    'Regulation' = list(
        'GO:0010507' = 'negative',
        'GO:0010508' = 'positive'
    ),
    'Selective' = list(
        'GO:0035973' = 'aggrephagy',
        'GO:0000423' = 'mitophagy',
        'GO:0061709' = 'reticulophagy',
        'GO:0098792' = 'xenophagy'
    ),
    'Subtypes' = list(
        'GO:0000422' = 'mitochondrion',
        'GO:0030242' = 'peroxisome',
        'GO:0061684' = 'chaperone-mediated',
        'GO:0061738' = 'late-endosomal',
        'GO:0016236' = 'macroautophagy')
) %>%
    melt() %>%
    setNames(c('subcat', 'pathway', 'cat'))

gene_set_enrichment <- read_rds('data/gene_set_enrichment.rds')

nes <- gene_set_enrichment %>%
    filter(grepl('None', group)) %>%
    right_join(cats) %>%
    separate(group, c('course', 'time'), '_') %>%
    mutate(time = fct_reorder(factor(time), as.numeric(time))) %>%
    dcast(cat + subcat + pathway ~ time, value.var = 'NES') 

png(filename = 'manuscript2/figures/autophagy_categories.png',
    height = 8, width = 10, units = 'cm', res = 300)

Heatmap(as.matrix(nes[, 4:ncol(nes)]),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        row_split = nes$cat,
        row_labels = nes$subcat,
        column_title = 'Time (hr)',
        row_title = 'Enrichment Score',
        row_title_side = 'right',
        row_names_side = 'left',
        column_title_side = 'bottom',
        show_heatmap_legend = FALSE)

dev.off()
```

```{r factors_course}
# data prep ----
gene_annotations <- read_rds('data/gene_annotations.rds') %>%
    filter(grepl('Autophagy', category)) %>%
    dplyr::select(geneId = ENTREZID, gene_id = SYMBOL) %>%
    unique() %>%
    as_tibble()

res <- read_rds('data/differential_expression.rds')

binding_peaks <- read_rds('data/binding_peaks.rds')
targets <- mcols(binding_peaks) %>%
    as_tibble() %>%
    dplyr::select(factor_time = time, factor, geneId) %>%
    unique() %>%
    right_join(gene_annotations) %>%
    na.omit()

# figure ----
png(filename = 'manuscript2/figures/factors_course.png',
    height = 16, width = 16, units = 'cm', res = 300)

grid.newpage()

pushViewport(viewport(layout = grid.layout(nr = 2, nc = 2)))
pushViewport(viewport(layout.pos.row = 1, layout.pos.col = 1))

# A ----

pparg_target_expresss <- targets %>%
    left_join(res) %>%
    filter(factor == 'PPARG',
           course == 'None') %>%
    acast(factor_time ~ time,
          value.var = 'fold_change',
          fun.aggregate = median)
col_fun = circlize::colorRamp2(c(min(pparg_target_expresss), 0, max(pparg_target_expresss)), c("blue", "white", "red"))
draw(
    Heatmap(pparg_target_expresss,
            show_row_names = TRUE,
            cluster_columns = FALSE,
            row_dend_width = unit(7, "mm"),
            show_heatmap_legend = FALSE,
            row_title = 'Binding Time (hr)',
            row_title_side = 'right',
            column_title = 'Expression Time (hr)',
            column_title_side = 'bottom',
            col = col_fun),
  newpage = FALSE,
  padding = unit(c(5, 10, 10, 5), "mm")
)

grid.text("A", x = 0, y = 1, just = c("left", "top"),
          gp = gpar(fontsize = 14))
upViewport()

# B ----
pushViewport(viewport(layout.pos.row = 1, layout.pos.col = 2, just = 'left'))
cebpb_target_expresss <- targets %>%
    left_join(res) %>%
    filter(factor == 'CEBPB',
           course == 'None') %>%
    acast(factor_time ~ time,
          value.var = 'fold_change',
          fun.aggregate = median)
col_fun = circlize::colorRamp2(c(min(cebpb_target_expresss), 0, max(cebpb_target_expresss)), c("blue", "white", "red"))

draw(
  
    Heatmap(cebpb_target_expresss,
            show_row_names = TRUE,
            cluster_columns = FALSE,
            row_dend_width = unit(7, "mm"),
            show_heatmap_legend = FALSE,
            row_title = 'Binding Time (hr)',
            row_title_side = 'right',
            column_title = 'Expression Time (hr)',
            column_title_side = 'bottom',
            col = col_fun),
  newpage = FALSE,
  padding = unit(c(5, 10, 10, 5), "mm")
)

grid.text("B", x = 0, y = 1, just = c("left", "top"),
          gp = gpar(fontsize = 14))

upViewport()

# C ----
pushViewport(viewport(layout.pos.row = 2, just = 'center'))
enrich_category <- as.matrix(nes[, 4:ncol(nes)])
col_fun = circlize::colorRamp2(c(min(enrich_category), 0, max(enrich_category)), c("blue", "white", "red"))

draw(
  Heatmap(enrich_category,
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        row_split = nes$cat,
        row_labels = nes$subcat,
        column_title = 'Time (hr)',
        row_title = 'Enrichment Score',
        row_title_side = 'right',
        row_names_side = 'left',
        column_title_side = 'bottom',
        show_heatmap_legend = FALSE,
            rect_gp = gpar(type = "none"),
            cell_fun = function(j, i, x, y, width, height, fill) {
                r = 1.2*abs(enrich_category[i, j])/2 * min(unit.c(width, height))
                gp = gpar(fill = col_fun(enrich_category[i, j]), col = NA)
              grid.circle(x = x, y = y, r = r, gp = gp)
            }),
  newpage = FALSE,
  padding = unit(c(5, 45, 5, 5), "mm")
)

grid.text("C", x = 0, y = 1, just = c("left", "top"),
          gp = gpar(fontsize = 14))

upViewport()

dev.off()
```

```{r model_factors}
model_factors_short <- read_rds('data/model_factors_short.rds')
states_factor_short <- c('Insulated',
            'Insulated + Ubiquitous Binding', 'Ubiquitous Binding',
            'Co-Factor Binding',
            'Factor Devoid 1', 
            'Early Acting (CEBPB/MED1)',
            'Factor Devoid 2',
            'Factor Devoid 3',
            'Repressed',
            'Long term (PPARG/RXRG)')
short_emission <- emission(model_factors_short)
rownames(short_emission) <- states_factor_short
short_transition <- transition(model_factors_short)
rownames(short_transition) <- states_factor_short
colnames(short_transition) <- states_factor_short

model_factors_long <- read_rds('data/model_factors_long.rds')
states_factor_long <- c('CEBPB + (MED1/RXRG)',
            'Factor Devoid 1', 
            'Long term (PPARG/RXRG)',
            'Factor Devoid 2', 
            'Long term (PPARG)',
            'PPARG + (RXRG/MED1)',
            'Ubiquitous Binding',
            'Insulated')

long_emission <- emission(model_factors_long)
rownames(long_emission) <- states_factor_long
long_transition <- transition(model_factors_long)
rownames(long_transition) <- states_factor_long
colnames(long_transition) <- states_factor_long

model <- list(
    list('Emission Probability' = short_emission,
         'Transition Probability' = short_transition),
    list('Emission Probability' = long_emission,
        'Transition Probability' = long_transition)
)
names(model) <- c('short', 'long')

# figure
png(filename = 'manuscript2/figures/model_factors.png',
    height = 12, width = 32, units = 'cm', res = 300)

grid.newpage()
pushViewport(viewport(layout = grid.layout(nr = 1, nc = 2)))
pushViewport(viewport(layout.pos.row = 1, layout.pos.col = 1))
col_func <- circlize::colorRamp2(c(0, 1), c('white', 'black'))

draw(
  Heatmap(model$short$`Emission Probability`,
        col = col_func,
        show_heatmap_legend = FALSE,
        show_row_dend = FALSE,
        show_column_dend = FALSE,
        row_names_side = 'left',
        column_names_rot = 45,
        column_title = 'Emissions') +
  Heatmap(model$short$`Transition Probability`,
        col = col_func,
        cluster_columns = FALSE,
        show_heatmap_legend = FALSE,
        show_row_dend = FALSE,
        show_column_dend = FALSE,
        row_names_side = 'left',
        column_names_rot = 45,
        column_title = 'Transitions'),
  newpage = FALSE
)
grid.text("A", x = 0, y = 1, just = c("left", "top"),
          gp = gpar(fontsize = 18))
upViewport()

pushViewport(viewport(layout.pos.row = 1, layout.pos.col = 2))

draw(
  Heatmap(model$long$`Emission Probability`,
        col = col_func,
        cluster_columns = FALSE,
        show_heatmap_legend = FALSE,
        show_row_dend = FALSE,
        show_column_dend = FALSE,
        row_names_side = 'left',
        column_names_rot = 45,
        column_title = 'Emissions') +
  Heatmap(model$long$`Transition Probability`,
        col = col_func,
        show_heatmap_legend = FALSE,
        show_row_dend = FALSE,
        show_column_dend = FALSE,
        row_names_side = 'left',
        column_names_rot = 45,
        column_title = 'Transitions'),
  newpage = FALSE
)
grid.text("B", x = 0, y = 1, just = c("left", "top"),
          gp = gpar(fontsize = 18))

upViewport()

dev.off()
```

```{r state_frequencies}
gene_annotations <- read_rds('data/gene_annotations.rds') %>%
    dplyr::select(category, geneId = ENTREZID) %>%
    unique()

plot_grid(
    map_df(segment(model_factors_short), as_tibble, .id = 'time') %>%
    left_join(gene_annotations) %>%
    left_join(tibble(state = paste0('E', 1:10),
                     name = states_factor_short)) %>%
    filter(category %in% c('Autophagy', 'Lipogenesis'),
           abs(distanceToTSS) < 50000,
           name %in% states_factor_short[c(1, 3, 4, 9)]) %>%
    mutate(time = str_split(time, '_', simplify = TRUE)[, 2]) %>%
    ggplot(aes(x = name, fill = time)) +
    geom_bar(position = 'dodge', width = .8) +
    coord_flip() +
    labs(y = 'State Frequency', x = '', fill = 'Time (hr)') +
    facet_wrap(~category, scales = 'free_x', nrow = ) +
    scale_fill_manual(values = c('black', 'darkgray')) +
    scale_x_discrete(labels = wrap_format(10)) +
    theme(legend.position = c(.83, .91),
          legend.direction = 'horizontal',
          legend.background = element_blank(),
          panel.grid = element_blank(),
          strip.background = element_blank(),
          strip.text = element_text(size = 12),
          axis.text = element_text(size = 12),
          axis.title = element_text(size = 14),
          plot.margin = unit(c(0,0,0,0), "cm")),
    map_df(segment(model_factors_long), as_tibble, .id = 'time') %>%
    left_join(gene_annotations) %>%
    left_join(tibble(state = paste0('E', 1:8),
                     name = states_factor_long)) %>%
    filter(category %in% c('Autophagy', 'Lipogenesis'),
           abs(distanceToTSS) < 50000,
           name %in% states_factor_long[c(1, 3, 5, 6)]) %>%
    mutate(time = str_split(time, '_',simplify = TRUE)[, 2],
           time = fct_reorder(time, as.numeric(time))) %>%
    ggplot(aes(x = time, fill = name)) +
    geom_bar(aes(y = (..count..)/sum(..count..)), width = .8) + 
    facet_wrap(~category, scales = 'free_y') +
    labs(x = 'Time (hr)', y = 'State Frequency', fill = 'States') +
    theme(legend.direction = 'vertical',
          panel.grid = element_blank(),
          strip.background = element_blank(),
          strip.text = element_text(size = 12),
          axis.text = element_text(size = 10),
          axis.title = element_text(size = 14),
          plot.margin = unit(c(0,0,0,0), "cm"),
          legend.key.size = unit(.8,"line")),
scale = .95,
ncol = 1,
labels = 'AUTO',
label_fontface = 'plain'
) %>%
    ggsave(plot = .,
           filename = 'manuscript2/figures/state_frequencies.png',
           height = 7, width = 7)
```

```{r binding_overlap}
binding_peaks <- read_rds('data/binding_peaks.rds')
gene_annotations <- read_rds('data/gene_annotations.rds') %>%
    with(split(ENTREZID, category)) %>%
    map(unique)

peaks_by_group <-  mcols(binding_peaks) %>%
    as_tibble() %>%
    filter(!factor %in% c('CTCF', 'EP300'),
           geneId %in% gene_annotations$Autophagy) %>%
    group_by(factor, geneId) %>%
    summarise(times = list(unique(time))) %>%
    ungroup() %>%
    group_split(factor)

over_representation <- read_rds('data/targets_over_representation.rds')

names(peaks_by_group) <- unlist(map(peaks_by_group, function(x) unique(x$factor)))

plot_grid(
    plot_grid(
  peaks_by_group$PPARG %>%
  ggplot(aes(x = times)) +
  geom_bar(width = .8) +
  scale_x_upset(n_intersections = 7) +
  labs(x = '',
       y = 'Number of Targets') +
  ggtitle('PPARG') +
  theme(panel.grid = element_blank(),
        axis.ticks.x = element_blank(),
        plot.title = element_text(hjust = .95, margin = margin(t = 0, b = -20, l = 0, r = 0)),
        plot.background = element_blank(),
        axis.title = element_text(size = 14)),
peaks_by_group$RXRG %>%
  ggplot(aes(x = times)) +
  geom_bar(width = .8) +
  scale_x_upset(n_intersections = 7) +
  labs(x = '',
       y = '') +
  ggtitle('RXRG') +
  lims(y = c(0, 15)) +
  theme(panel.grid = element_blank(),
        axis.ticks.x = element_blank(),
        plot.title = element_text(hjust = .95, margin = margin(t = 0, b = -20, l = 0, r = 0)),
        plot.background = element_blank()),
peaks_by_group$CEBPB %>%
  ggplot(aes(x = times)) +
  geom_bar(width = .8) +
  scale_x_upset(n_intersections = 7) +
  labs(x = 'Intersections',
       y = 'Number of Targets') +
  ggtitle('CEBPB') +
  theme(panel.grid = element_blank(),
        axis.ticks.x = element_blank(),
        plot.title = element_text(hjust = .95, margin = margin(t = 0, b = -20, l = 0, r = 0)),
        plot.background = element_blank(),
        axis.title = element_text(size = 14),),
peaks_by_group$MED1 %>%
  ggplot(aes(x = times)) +
  geom_bar(width = .8) +
  scale_x_upset(n_intersections = 7) +
  labs(x = 'Intersections',
       y = '') +
  ggtitle('MED1') +
  theme(panel.grid = element_blank(),
        axis.ticks.x = element_blank(),
        plot.title = element_text(hjust = .95, margin = margin(t = 0, b = -20, l = 0, r = 0)),
        axis.title = element_text(size = 14),
        plot.background = element_blank()),
nrow = 2,
scale = 1.08
),
    as_tibble(over_representation) %>% 
    separate(group, into = c('subject', 'subject_time')) %>%
    separate(ID, into = c('query', 'query_time')) %>%
    separate(GeneRatio, into = c('count', 'bg'), sep = '/') %>%
    mutate(ratio = as.numeric(count)/as.numeric(bg),
           subject_time = fct_reorder(factor(subject_time),
                                      as.numeric(subject_time))) %>%
    filter(category == 'Autophagy',
           subject_time == query_time,
           subject %in% c('MED1', 'RXRG', 'EP300'),
           query %in% c('PPARG', 'CEBPB')) %>%
    ggplot(aes(x = subject_time, y = ratio)) +
    geom_col(width = .8) +
    facet_grid(query~subject, scales = 'free_x', space = 'free_x') +
    labs(x = 'Time (hr)', y = 'Fraction of Overlapping Targets') +
    theme(panel.grid = element_blank(),
          strip.background = element_blank(),
          strip.text = element_text(size = 14),
          axis.text = element_text(size = 12),
          axis.title = element_text(size = 14),
          plot.margin = unit(c(0,0,0,0), 'mm')),
    scale = .9,
    nrow = 1,
    labels = 'AUTO',
    label_fontface = 'plain'
) %>%
    ggsave(plot = .,
           filename = 'manuscript2/figures/binding_overlap.png',
           height = 6, width = 12)
```

```{r peaks_overlap_enrichment}
# hm model
model_hm <- read_rds('data/model_hm.rds')
states <- c('Active Promoter', 'Weak Enhancer', 'Active Enhancer',
            'Genic Enhancer 1', 'Genic Enhancer 2', 'Heterochromatin ',
            'Repeats', 'Repressed Polycomb', 'Strong Transcription')
hm_emission <- emission(model_hm)
rownames(hm_emission) <- states
hm_transition <- transition(model_hm)
rownames(hm_transition) <- states
colnames(hm_transition) <- states

# peaks overlap enrichment
peaks_overlap_enrichment <- read_rds('data/peaks_overlap_enrichment.rds')
mats <- peaks_overlap_enrichment %>%
    group_by(hour) %>%
    group_split() %>%
    map(function(x) {
        m <- dplyr::select(x, -hour, -state, -genome) %>% as.matrix()
        nms <- str_split(colnames(m), '_', simplify = TRUE)
        ind <- nms[, 2] == unique(x$hour)
        m <- (m[, ind])
        rownames(m) <- states
        colnames(m) <- nms[, 1][ind]
        m
    })
names(mats) <- unique(peaks_overlap_enrichment$hour)

png(filename = 'manuscript2/figures/peaks_overlap_enrichment.png',
    height = 9, width = 28, units = 'cm', res = 300)

grid.newpage()
pushViewport(viewport(layout = grid.layout(nr = 1, nc = 2)))
pushViewport(viewport(layout.pos.row = 1, layout.pos.col = 1))
col_func <- circlize::colorRamp2(c(0, 1), c('white', 'black'))


draw(
  Heatmap(hm_emission,
        col = col_func,
        show_heatmap_legend = FALSE,
        show_row_dend = FALSE,
        show_column_dend = FALSE,
        row_names_side = 'left',
        column_names_rot = 45,
        column_title = 'Emissions') +
  Heatmap(hm_transition,
        col = col_func,
        cluster_columns = FALSE,
        show_heatmap_legend = FALSE,
        show_row_dend = FALSE,
        show_column_dend = FALSE,
        row_names_side = 'left',
        column_names_rot = 45,
        column_title = 'Transitions'),
  newpage = FALSE
)
grid.text("A", x = 0, y = 1, just = c("left", "top"),
          gp = gpar(fontsize = 18))
upViewport()

pushViewport(viewport(layout.pos.row = 1, layout.pos.col = 2))

col_func <- circlize::colorRamp2(c(0, 56), c('white', 'black'))

draw(
  Heatmap(mats$`0`,
          col = col_func,
          show_row_dend = FALSE,
          show_column_dend = FALSE,
          row_names_side = 'left',
          column_names_rot = 45,
          show_heatmap_legend = FALSE,
          column_title = '0 hour') +
  Heatmap(mats$`48`,
          col = col_func,
          show_column_dend = FALSE,
          row_names_side = 'left',
          column_names_rot = 45,
          show_heatmap_legend = FALSE,
          column_title = '48 hours')+
  Heatmap(mats$`168`,
          col = col_func,
          show_column_dend = FALSE,
          row_names_side = 'left',
          column_names_rot = 45,
          show_heatmap_legend = FALSE,
          column_title = '148 hours'),
  newpage = FALSE
)
grid.text("B", x = 0, y = 1, just = c("left", "top"),
          gp = gpar(fontsize = 18))

upViewport()

dev.off()
```

```{r molecular_functions}
mf_overrepresentation <- read_rds('data/molecular_functions_overrepresentation.rds')

# terms <- c('transcription factor binding',
#            'transcription coactivator activity',
#            'enzyme regulator activity',
#            'transcription regulator activity',
#            'transmembrane transporter activity',
#            'transcription coregulator activity')

terms <- c('protein kinase activity',
           'DNA-binding transcription factor activity',
           # 'ubiquitin binding',
           # 'oxidoreductase activity',
           # 'hydrolase activity',
           # 'transferase activity'
           'phosphatase activity')

terms_short <- c('kinases', 'transcription factors', 'phosphatase')

gene_annotations <- read_rds('data/gene_annotations.rds') %>%
    dplyr::select(gene_id = ENTREZID, symbol = SYMBOL)

mf_annotations <- read_rds('data/mf_annotations.rds') %>%
    filter(term %in% terms) %>%
    inner_join(gene_annotations) %>%
    dplyr::select(category, term, gene_id = symbol) %>%
    unique()

differential_expression <- read_rds('data/differential_expression.rds')
plot_grid((mf_overrepresentation %>%
        as_tibble() %>%
        separate(group, c('factor', 'time')) %>%
        separate(GeneRatio, c('count', 'bg')) %>%
        mutate(ratio = as.numeric(count)/as.numeric(bg),
               time = fct_reorder(factor(time), as.numeric(time))) %>%
        filter(ID %in% terms,
               factor %in% c('CEBPB', 'PPARG')) %>%
        left_join(tibble(ID = terms,
                         short = terms_short)) %>%
        ggplot(aes(x = time, y = ratio, fill = short)) +
        geom_col(position = 'dodge', width = .8) +
        facet_grid(category ~ factor, scales = 'free', drop = FALSE) +
        theme(legend.position = c(.2, .9),
              legend.direction = 'vertical',
              legend.background = element_blank(),
              panel.grid = element_blank(),
              strip.background = element_blank(),
              strip.text = element_text(size = 14),
              axis.text = element_text(size = 10),
              axis.title = element_text(size = 14),
              plot.margin = unit(c(0,0,0,0), 'mm')) +
        labs(x = 'Time (hr)', y = 'Fraction of Targets', fill = '')),
  
(differential_expression %>%
        inner_join(mf_annotations) %>%
        left_join(tibble(term = terms, short = terms_short)) %>%
        filter(category == 'Autophagy') %>%
        mutate(time = fct_reorder(factor(time), as.numeric(time))) %>%
        ggplot(aes(x = time, y = (fold_change))) +
        geom_boxplot() +
        facet_grid(short~course, scales = 'free_x', space = 'free_x') +
        theme(panel.grid = element_blank(),
              strip.background = element_blank(),
              strip.text = element_text(size = 12),
              axis.text = element_text(size = 10),
              axis.title = element_text(size = 14),
              plot.margin = unit(c(0,0,0,0), 'mm')) +
        labs(x = 'Time (hr)', y = 'Fold-change (log_2)', fill = '')),
scale = .95, 
labels = 'AUTO',
label_fontface = 'plain') %>%
    ggsave(plot = .,
           filename = 'manuscript2/figures/molecular_functions.png',
           height = 5.5, width = 10)
```

```{r expression_data}
adipo_counts <- read_rds('data/adipo_counts.rds')
pd <- colData(adipo_counts)[adipo_counts$time %in% c(0, 2, 4, 24, 48, 144, 168, 192, 240),] %>%
    as_tibble() %>%
    dplyr::select(study_name, time, bibtexkey)

pd %>%
    group_by(study_name, bibtexkey) %>%
    summarise(n = n(),
              time = paste(unique(time), collapse = '/')) %>%
    ungroup() %>%
    dplyr::select(study_name, n, time, bibtexkey) %>%
    mutate(bibtexkey = paste0('\\cite{', bibtexkey, '}')) %>%
    setNames(c('GEO ID', 'N', 'Time (hr)', 'Ref.')) %>%
    xtable(align = 'clclc') %>%
    print(floating = FALSE,
          include.rownames = FALSE,
          booktabs = TRUE,
          sanitize.text.function = identity,
          comment = FALSE,
          file = 'manuscript2/tables/expression_data.tex')
```

```{r binding_data}
md <- map_df(
    c('data/segmentation/input_files_factors_long.tsv',
      'data/segmentation/input_files_factors_short.tsv'),
    read_tsv,
    col_names = c('time', 'factor', 'gsm')
) %>%
    mutate(gsm = str_split(gsm, '\\.', simplify = TRUE)[, 1])
 peak_counts <- read_rds('data/peak_counts.rds')
pd <- colData(peak_counts) %>% 
    as_tibble() %>%
    filter(id %in% md$gsm) %>%
    dplyr::select(study, factor, bibtexkey)

pd %>%
    group_by(study, bibtexkey) %>%
    summarise(n = n(),
              factor = paste(unique(factor), collapse = '/ ' )) %>%
    ungroup() %>%
    dplyr::select(study, n, factor, bibtexkey) %>%
    mutate(bibtexkey = paste0('\\cite{', bibtexkey, '}')) %>%
    setNames(c('SRA ID', 'N', 'Antibody', 'Ref.')) %>%
    xtable(align = 'clcp{.15\\textwidth}c') %>%
    print(floating = FALSE,
          include.rownames = FALSE,
          booktabs = TRUE,
          sanitize.text.function = identity,
          comment = FALSE,
          file = 'manuscript2/tables/binding_data.tex')
```

```{r hm_data}
md <- read_tsv('data/segmentation/hm/input_files_hm.tsv',
               col_names = c('time', 'hm', 'gsm')) %>%
    mutate(gsm = str_split(gsm, '\\.', simplify = TRUE)[, 1])
# peak_counts <- read_rds('data/peak_counts.rds')
pd <- colData(peak_counts) %>% 
    as_tibble() %>%
    filter(id %in% md$gsm) %>%
    dplyr::select(study, factor, bibtexkey)

pd %>%
    group_by(study, bibtexkey) %>%
    summarise(n = n(),
              factor = paste(unique(factor), collapse = '/ ')) %>%
    ungroup() %>%
    dplyr::select(study, n, factor, bibtexkey) %>%
    mutate(bibtexkey = paste0('\\cite{', bibtexkey, '}')) %>%
    setNames(c('SRA ID', 'N', 'Antibody', 'Ref.')) %>%
    xtable(align = 'clcp{.15\\textwidth}c') %>%
    print(floating = FALSE,
          include.rownames = FALSE,
          booktabs = TRUE,
          sanitize.text.function = identity,
          comment = FALSE,
          file = 'manuscript2/tables/hm_data.tex')
```

